### PR TITLE
GH-106897: Add `RERAISE` event to `sys.monitoring`.

### PR DESCRIPTION
**Pull request by markshannon**
Fixes the imbalance between `RAISE` and `EXCEPTION_HANDLED` events.

* Issue: gh-106897

### CODE REVIEW

**Comment by markshannon** [comment]
@gaogaotiantian would you mind reviewing this?

**Comment by gaogaotiantian** [comment]
If I understand the monitoring system correctly, for all the exceptions raised, we want to balance it so each `RAISE` or `RERAISE` event(except for `StopIteration`) should match a `PY_UNWIND` or an `EXCEPTION_HANDLED` event right?

Consider the following code:

```python
import sys
import asyncio

E = sys.monitoring.events

sys.monitoring.use_tool_id(0, "test")
 
async def async_generator():
    for i in range(1):
        raise ZeroDivisionError
        yield i
 
async def main():
    sys.monitoring.register_callback(0, E.EXCEPTION_HANDLED, lambda *args: print(f"EXCEPTION_HANDLED {args}"))
    sys.monitoring.register_callback(0, E.RAISE, lambda *args: print(f"RAISE {args}"))
    sys.monitoring.register_callback(0, E.RERAISE, lambda *args: print(f"RERAISE {args}"))
    sys.monitoring.register_callback(0, E.PY_UNWIND, lambda *args: print(f"PY_UNWIND {args}"))
    sys.monitoring.set_events(0, E.EXCEPTION_HANDLED | E.RAISE | E.RERAISE | E.PY_UNWIND)
    try:
        async for item in async_generator():
            print(item)
    except Exception:
        pass
    sys.monitoring.set_events(0, 0)
 
asyncio.run(main())
```

It prints

```
RAISE (<code object async_generator at 0x7f54aaeb5230, file "/home/gaogaotiantian/programs/cpython/example.py", line 10>, 44, ZeroDivisionError())
EXCEPTION_HANDLED (<code object async_generator at 0x7f54aaeb5230, file "/home/gaogaotiantian/programs/cpython/example.py", line 10>, 50, ZeroDivisionError())
RERAISE (<code object async_generator at 0x7f54aaeb5230, file "/home/gaogaotiantian/programs/cpython/example.py", line 10>, 52, ZeroDivisionError())
PY_UNWIND (<code object async_generator at 0x7f54aaeb5230, file "/home/gaogaotiantian/programs/cpython/example.py", line 10>, 52)
RAISE (<code object main at 0x5600748fdbc0, file "/home/gaogaotiantian/programs/cpython/example.py", line 17>, 610, ZeroDivisionError())
EXCEPTION_HANDLED (<code object main at 0x5600748fdbc0, file "/home/gaogaotiantian/programs/cpython/example.py", line 17>, 656, ZeroDivisionError())
EXCEPTION_HANDLED (<code object main at 0x5600748fdbc0, file "/home/gaogaotiantian/programs/cpython/example.py", line 17>, 660, ZeroDivisionError())
```

Maybe because we should add `monitor_reraise` (or maybe `monitor_raise`?) in `END_ASYNC_FOR`? The exception from the coroutine seems to be raised again.

**Comment by markshannon** [comment]
Thanks for spotting that.

File: Python/bytecodes.c [hunk 1]
```diff
@@ -720,7 +720,11 @@ dummy_func(
                 exc = args[0];
                 /* fall through */
             case 0:
-                ERROR_IF(do_raise(tstate, exc, cause), exception_unwind);
+                if (do_raise(tstate, exc, cause)) {
```
Comment (iritkatriel) [inline Python/bytecodes.c:723]: I'm confused about error reporting from do_raise. If there is no exception set, it sets an exception and returns 0. What does that mean?

Comment (markshannon) [reply Python/bytecodes.c:723]: `do_raise` returns 0 if it raises, and 1 if it **re**raises.
Historical reasons probably :shrug:

**Review by iritkatriel — APPROVED** [review]

**Pull request merged**